### PR TITLE
Stabilize Asynchronous Expectations

### PR DESCRIPTION
**Pull request by jeffh**
Make async expectations less flaky.

This commit utilizes dispatch sources to perform all its asynchronous
polling / timeout behavior instead of relying on the run loop. Sources
is a lower-level than using NSRunLoop which gives us more control over
when timers and polling intervals are executed.

Also, nesting any combination of waitUntil and
expect(...).toEventually(...) **is not** allowed because all asynchronous
checking code manages the run loop, which makes it inheritantly
unreliable to use at the same time. Nimble will now throw an exception
when it detects this scenario, although it is up to the test writer to
ensure he/she doesn't manually manipulate the run loop while
asynchronous expectations are on-going.

~~Breaking change: waitUntil's closure argument is executed synchronously
instead of dispatching to asynchronously to the main thread.~~

-----------------------------

TODOs:

- [x] See if `PollResult` can be removed / merged with `AwaitPromise`
- [x] Repeatedly run to make sure it's actually more stable
- [x] Refactor Await* classes
  - Avoid globals - have an `Environment` class to manage "global" state
- [x] See if waitUntil's argument can be dispatched async to main thread for backwards compatibility
- [x] Get rid of the Logging.swift file
- [x] Remove dead codes from exploration
- [x] Expand "stalled" error message to explain more details (eg - XCTest's expectations error)
- [x] Documentation

Related Issues:

- Raises error for cases described in #223 
- #216
- #180
- #177

### CODE REVIEW

**Comment by tmandry** [comment]
Curious what you mean about the new waitUntil behavior.. does that mean it will only wait for changes that are caused by background threads?

**Comment by jeffh** [comment]
This PR's `waitUntil` implementation is more similar to XCTest's expectations:

```swift
func xctest() {
    let expect = expectationWithDescription(...)
    // do async code
    dispatch_async(dispatch_get_main_queue()) {
       expect.fulfill()
    }
    // end
    waitForExpectations(...)
}
// equal to above
func nimble() {
    waitUntil { done in
        // do async code
        dispatch_async(dispatch_get_main_queue()) {
           done()
        }
        // end
    }
}
```

In the current implementation of `waitUntil`, it dispatches the closure to main thread asynchronously:

```swift
func nimble() {
    waitUntil { done in
       done() // dispatch asynchronously to main thread
    }
}
```

Unfortunately, doing this implicit dispatch currently conflicts with trying to provide helpful error messages for nested async expectations. It needs to be resolved before merging this PR (or remove the helpful error message).

**Comment by jeffh** [comment]
This PR is ready for feedback, @Quick/contributors.

File: NimbleTests/AsynchronousTest.swift [hunk 1]
```diff
@@ -63,22 +76,77 @@ class AsyncTest: XCTestCase {
...
+    }
 
-        // "clear" runloop to ensure this test doesn't poison other tests
-        NSRunLoop.mainRunLoop().runUntilDate(NSDate().dateByAddingTimeInterval(2.0))
+    func testWaitUntilErrorsIfDoneIsCalledMultipleTimes() {
+        waitUntil { done in
+            deferToMainQueue {
+                done()
+            }
+        }
```
Comment (ashfurrow) [inline NimbleTests/AsynchronousTest.swift]: Is this invocation of `waitUntil` meant to be part of this test case?

File: NimbleTests/AsynchronousTest.swift [hunk 2]
```diff
@@ -63,22 +76,77 @@ class AsyncTest: XCTestCase {
...
+        let referenceLine = __LINE__ + 9
+        var msg = "Unexpected exception raised: Nested async expectations are not allowed "
+        msg += "to avoid creating flaky tests."
+        msg += "\n\n"
+        msg += "The call to\n\t"
+        msg += "expect(...).toEventually(...) at \(__FILE__):\(referenceLine + 7)\n"
+        msg += "triggered this exception because\n\t"
+        msg += "waitUntil(...) at \(__FILE__):\(referenceLine + 1)\n"
+        msg += "is currently managing the main run loop."
+        failsWithErrorMessage(msg) { // reference line
```
Comment (ashfurrow) [inline NimbleTests/AsynchronousTest.swift]: `reference line`?

File: README.md [hunk 3]
```diff
@@ -373,6 +379,12 @@ waitUntilTimeout(10, ^(void (^done)(void)){
 });
 ```
 
+Note: waitUntil triggers its timeout code on the main thread. Blocking the main
```
Comment (ashfurrow) [inline README.md]: :+1: for documentation :tada:

**Comment by ashfurrow** [comment]
I'm only somewhat familiar with the Nimble codebase, but everything here makes sense to me :ok_hand:

File: NimbleTests/AsynchronousTest.swift (hunk 2 above)
Comment (jeffh) [inline NimbleTests/AsynchronousTest.swift]: It's just a line number to make all the other line number indicators derived of. This makes it a bit easier to change the error message without having to update every time the error message changes.

The error message emits line numbers based on where `waitUntil` and `toEventually` is used.

**Comment by jeffh** [comment]
Thanks for the review @ashfurrow!

**Pull request merged**